dnsdist: Fix short IOs over TCP

rgacogne opened the pull request:
### Short description
<!-- Write a small description of what this Pull Request fixes or provides, including the issue #s -->
`TCPIOHandler::tryRead()` and `TCPIOHandler::tryWrite()` are supposed to update the position and read/write the remaining number of bytes based on the initial total number minus the position, so that the caller can just pass the exact same parameters on short reads/writes until it succeeds or an exception is raised, without having to care about updating the position and number of bytes to read/write.
<strike>Untested yet.</strike>

Supersedes #7971.

### Checklist
<!-- please indicate if any of these things are done/included with this Pull Request. Not all boxes need to be checked for the Pull Request to be accepted -->
I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [ ] included documentation (including possible behaviour changes)
- [ ] documented the code
- [x] added or modified regression test(s)
- [ ] added or modified unit test(s)

omoerbeek commented on pdns/tcpiohandler.hh:
```diff
@@ -208,9 +208,8 @@ public:
       return d_conn->tryRead(buffer, pos, toRead);
     }
 
```
The cases `toRead == pos` (will cause `::read` to rerturn `0`) and `pos > toRead` (sign wrap) should be handled.

omoerbeek commented on pdns/tcpiohandler.hh:
```diff
@@ -243,9 +241,8 @@ public:
       return d_conn->tryWrite(buffer, pos, toWrite);
     }
 
```
same here

rgacogne commented:
Added some checks suggested by @omoerbeek and fixed the existing one, which compared the size of the buffer to the requested size, not the requested size minus the existing position.

Habbie commented:
>  Untested yet.

It fixes the problem in my test environment!

rgacogne commented:
I forgot to update my comment but I actually tested it in the meantime! I reproduced the issues with short reads and short writes with master, and confirmed that this PR fixed them.
I'll try to write a test for these cases, since it's already the second time I manage to introduce a bug in that exact spot.

omoerbeek commented:
I approved this, but now it looks to me like not all cases of tryRead were covered.

cmouse commented on pdns/tcpiohandler.hh:
```diff
@@ -224,28 +223,29 @@ public:
-    size_t sent = 0;
     do {
-      ssize_t res = ::write(d_socket, reinterpret_cast<char*>(&buffer.at(pos)), toWrite - sent);
+      ssize_t res = ::write(d_socket, reinterpret_cast<char*>(&buffer.at(pos)), toWrite - pos);
```
are you accidentically removing constness here?

cmouse replied:
also, what if res < 0?

omoerbeek replied:
res < 0 case handled below

cmouse replied:
uhm, how?

omoerbeek replied:
line 252

cmouse replied:
*sigh*, i got fooled by the review view.

cmouse commented on pdns/tcpiohandler.hh:
```diff
@@ -192,25 +192,24 @@ public:
-    size_t got = 0;
     do {
-      ssize_t res = ::read(d_socket, reinterpret_cast<char*>(&buffer.at(pos)), toRead - got);
+      ssize_t res = ::read(d_socket, reinterpret_cast<char*>(&buffer.at(pos)), toRead - pos);
```
what if res < 0?

rgacogne commented:
> Seems a few calls to `tryRead` were missed when converting the 3rd arg. There are 4 calls here,
> line 1045 got converted, 1067 looks right, but 936 and 946 need fixing.

You are mistaking `TCPIOHandler::tryRead()` and a static function named `tryRead()`, which exists because we don't speak DoT toward backends atm and so have no need for the filters abstraction. I have a WIP branch unifying those two.

omoerbeek commented:
> > Seems a few calls to `tryRead` were missed when converting the 3rd arg. There are 4 calls here,
> > line 1045 got converted, 1067 looks right, but 936 and 946 need fixing.
> 
> You are mistaking `TCPIOHandler::tryRead()` and a static function named `tryRead()`, which exists because we don't speak DoT toward backends atm and so have no need for the filters abstraction. I have a WIP branch unifying those two.

Ah, right. Did not spot that and jumped to conlcusions. Having two conventions for the third arg *is* confusing...

rgacogne commented:
Now with regression tests.
